File: src/sceneio/io/_registry/families/dense.py
"""Built-in dense-MVS codec definitions."""

from __future__ import annotations

from pathlib import Path

from sceneio import _core
from sceneio.io._inspectors.dense import (
    inspect_colmap_fused_visibility,
    inspect_colmap_mvs_consistency,
    inspect_colmap_mvs_depth,
    inspect_colmap_mvs_normal,
)
from sceneio.io._registry.adapters import (
    _file_sink_writer,
    _mmap_reader,
    _mmap_selector_reader,
)
from sceneio.io._registry.model import Codec


def _inspector(function, payload_kind: str):
    def inspect(path: str):
        return function(Path(path), payload_kind)

    return inspect


DENSE_CODECS: tuple[Codec, ...] = (
    Codec(
        "colmap_mvs_depth",
        (),
        _mmap_reader(_core.read_colmap_mvs_depth),
        _file_sink_writer(_core.write_colmap_mvs_depth),
        record=_core.DepthMap,
        payload_kind="depth_map",
        inspect=_inspector(inspect_colmap_mvs_depth, "depth_map"),
        read_window=_mmap_selector_reader(
            _core.read_colmap_mvs_depth_window
        ),
        supported_features=(
            "scalar_float32",
            "little_endian",
            "nonpositive_invalid",
            "pixel_windows",
            "camera_z",
        ),
        unsupported_features=("confidence", "embedded_scale"),
    ),
    Codec(
        "colmap_mvs_normal",
        (),
        _mmap_reader(_core.read_colmap_mvs_normal),
        _file_sink_writer(_core.write_colmap_mvs_normal),
        record=_core.NormalMap,
        payload_kind="normal_map",
        inspect=_inspector(inspect_colmap_mvs_normal, "normal_map"),
        read_window=_mmap_selector_reader(
            _core.read_colmap_mvs_normal_window
        ),
        supported_features=(
            "planar_float32",
            "little_endian",
            "camera_xyz",
            "pixel_windows",
        ),
    ),
    Codec(
        "colmap_mvs_consistency",
        (),
        _mmap_reader(_core.read_colmap_mvs_consistency),
        _file_sink_writer(_core.write_colmap_mvs_consistency),
        record=_core.ConsistencyGraph,
        payload_kind="consistency_graph",
        inspect=_inspector(
            inspect_colmap_mvs_consistency, "consistency_graph"
        ),
        supported_features=(
            "ordered_pixel_entries",
            "zero_count_entries",
            "mvs_sequential_image_indices",
        ),
    ),
    Codec(
        "colmap_fused_visibility",
        (),
        _mmap_reader(_core.read_colmap_fused_visibility),
        _file_sink_writer(_core.write_colmap_fused_visibility),
        record=_core.PointVisibility,
        payload_kind="point_visibility",
        filenames=("fused.ply.vis",),
        inspect=_inspector(
            inspect_colmap_fused_visibility, "point_visibility"
        ),
        supported_features=(
            "csr",
            "zero_visibility_points",
            "mvs_sequential_image_indices",
        ),
    ),
)

File: src/cpp/bindings/dense.cpp
#include "bindings/registry.hpp"

void register_colmap_dense_mvs(nanobind::module_ &);

namespace sio::bindings {
namespace {

constexpr RegistrationDescriptor REGISTRATIONS[] = {
    {40, "colmap_dense_mvs", &::register_colmap_dense_mvs},
};

constexpr CodecDescriptor CODECS[] = {
    {56,
     "colmap_mvs_depth",
     "dense",
     symbols("read_colmap_mvs_depth"),
     symbols("write_colmap_mvs_depth"),
     symbols("_inspect_colmap_mvs_depth"),
     symbols("read_colmap_mvs_depth"),
     symbols("write_colmap_mvs_depth"),
     symbols("read_colmap_mvs_depth_window")},
    {57,
     "colmap_mvs_normal",
     "dense",
     symbols("read_colmap_mvs_normal"),
     symbols("write_colmap_mvs_normal"),
     symbols("_inspect_colmap_mvs_normal"),
     symbols("read_colmap_mvs_normal"),
     symbols("write_colmap_mvs_normal"),
     symbols("read_colmap_mvs_normal_window")},
    {58,
     "colmap_mvs_consistency",
     "dense",
     symbols("read_colmap_mvs_consistency"),
     symbols("write_colmap_mvs_consistency"),
     symbols("_inspect_colmap_mvs_consistency"),
     symbols("read_colmap_mvs_consistency"),
     symbols("write_colmap_mvs_consistency"),
     symbols()},
    {59,
     "colmap_fused_visibility",
     "dense",
     symbols("read_colmap_fused_visibility"),
     symbols("write_colmap_fused_visibility"),
     symbols("_inspect_colmap_fused_visibility"),
     symbols("read_colmap_fused_visibility"),
     symbols("write_colmap_fused_visibility"),
     symbols()},
};

constexpr FamilyBindings FAMILY{
    "dense",
    REGISTRATIONS,
    sizeof(REGISTRATIONS) / sizeof(REGISTRATIONS[0]),
    CODECS,
    sizeof(CODECS) / sizeof(CODECS[0]),
};

}  // namespace

const FamilyBindings &dense_bindings() { return FAMILY; }

}  // namespace sio::bindings
